File: 백준/Gold/17136. 색종이 붙이기/색종이 붙이기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
	/*
	 * 1x1, 2x2, 3x3, 4x4, 5x5 색종이가 5개씩 있음
	 * 0에는 색종이를 붙이면 안됨
	 * 1에다가 다 색종이를 붙여야함
	 * 1이 적힌 모든 칸을 붙이는데 필요한 색종이의 최소 갯수
	 * -------
	 * 그래프를 1만날때까지 순회함
	 * 1을 만나면 몇까지 붙일 수 있는지 체크해봄
	 * 체크후에 붙일 수 있는 모든 색종이들을 붙임
	 * 	이 과정에서 1을 붙였다면 1을 붙인상태에서 끝까지 가고 그다음 1을 다시 뗀다음 2를 붙여보고 하는식으로 가야함
	 * 다 돌았다면 다 0인지 체크 후 정답을 갱신해줌
	 * 
	 * 
	 */

	static int[][] graph;
	static int ans = Integer.MAX_VALUE;
	static int[] nums = {5,5,5,5,5};
	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		graph = new int[10][10];
		for(int i=0; i<10; i++) {
			st = new StringTokenizer(br.readLine());
			for(int j=0; j<10; j++) {
				graph[i][j] = Integer.parseInt(st.nextToken());
			}
		}
		
		DFS(0,0);
		
		if(ans==Integer.MAX_VALUE) System.out.println(-1);
		else System.out.println(ans);

	}
	
	static void DFS(int idx, int cnt) {
		if (cnt>=ans) return;
		
		if( idx==100) {
			ans = Math.min(ans, cnt);
			return;
		}
		
		int r = idx/10;
		int c = idx%10;
		
		if (graph[r][c]==0) {
			DFS(idx+1,cnt);
		}
		
		else {
			for(int range=0; range<5; range++) {
				if(nums[range]==0) continue;
				if(r+range>=10||c+range>=10) continue;
				if(!rangeChk(r, c, range)) continue;
				nums[range]--;
				put(r, c, range);
				DFS(idx+1, cnt+1);
				unPut(r, c, range);
				nums[range]++;
			}
			
		}
		
	}
	
	
	static boolean rangeChk(int r, int c, int range) {
		
		for(int i=r; i<=r+range; i++) {
			for(int j=c; j<=c+range; j++) {
				if(graph[i][j]==0) return false;
			}
		}
		return true;
	}
	
	static void put(int r, int c, int range) {
		
		for(int i=r; i<=r+range; i++) {
			for(int j=c; j<=c+range; j++) {
				graph[i][j] = 0;
			}
		}
		
		
	}
	
	static void unPut(int r, int c, int range) {
		for(int i=r; i<=r+range&&i<10; i++) {
			for(int j=c; j<=c+range&&j<10; j++) {
				graph[i][j] = 1;
			}
		}
		
	}

}

File: 백준/Gold/17136. 색종이 붙이기/색종이 붙이기.py
"""
입력:
  총 10x10에 종이의 상태들이 출력됨

출력:
  모든 1을 덮는데 필요한 색종이의 최소 개수
-------------------------------------
그리디로 풀기엔 반례가 있어서 DP로 풀어야 하는 문제

종이를 담아놓을 graph[10][10]
남아있는 색종이 갯수 paper[5] = [5,5,5,5,5]

1을 발견했을때 거기에 크기에 맞는 모든 종이들을 다 넣는다고 가정하자
(즉, 5x5크기의 1이면 처음 [0,0]에 크기1을 놓았을때와 크기5를 놓았을때까지 전부 재귀적으로 탐사)

이 함수의 종료조건은 모든 그래프를 다돌았을때(다 돌았다는 건 1이 하나도 없다는 의미를 뜻함)
재귀적으로 들어갈때 조건은 해당크기의 남은 색종이가 있을때, 10x10크기 안에서 인덱스가 존재할때
함수의 매개변수는 (y,x,cnt)
"""

import sys
input = sys.stdin.readline
sys.setrecursionlimit(100000)

graph = [list(map(int, input().split())) for _ in range(10)]

paper = [0 for _ in range(5)] #현재 쓴 색종이

result = 10000
def DFS(y, x, cnt):
  global result
  if y>=10:
    result = min(result, cnt)
    return
  elif x>=10: #if가 아닌 elif로 해줘야 밑에 graph[y][x] 에서 인덱스가 안터짐, 그리고 x>=10 이 넘어갈때 y+1 해주는거 잊지말기(안그럼 1행만 계속반복된다)
    DFS(y+1,0,cnt)

  elif graph[y][x]==1:

    for i in range(5): #최대 색종이크기가 5이므로 5까지 반복
      if y+i>=10 or x+i>=10: #인덱스가 터지면 넘어감
        continue
      if paper[i]>=5: #이미 쓴 색종이 갯수가 5개가 넘어가면 넘어감
        continue
      
      flag = 0 # 플래그 변수를 세워서 플래그가 변하는 순간 최대크기를 넘어간 것이므로 더 안들어감
      for j in range(i+1):  #1의 크기가 몇인지 일단 검사
        for k in range(i+1):
          if graph[y+j][x+k]==0:
            flag = 1
            break
        if flag ==1:break 
      
      if flag == 0:
        for j in range(i+1):
          for k in range(i+1):
            graph[y+j][x+k] = 0 #더 깊이 탐색할거니까 0으로 만들어줌
        paper[i]+=1 #색종이 한장 쓰고 들어감
        DFS(y, x+k, cnt+1) #y+j가 아닌 y임을 명심하자(y는 그대로여야 그 행의 남은x들을 탐색함)
        paper[i] -=1 #재귀에서 빠져나오는 순간 색종이 한장을 다시 빼주고
        
        for j in range(i+1):
          for k in range(i+1):
            graph[y+j][x+k] = 1 #그래프도 원상복구 시켜줌
  else:
    DFS(y,x+1,cnt)

DFS(0,0,0)
print(-1 if result==10000 else result)




